source/common/stats/thread_local_store.h

```diff
@@ -16,6 +16,142 @@
+  }
+  // Stats::Histogram
+  void recordValue(uint64_t value) override {
+    std::unique_lock<std::mutex> lock(merge_lock_);
+    hist_insert_intscale(histograms_[current_active_], value, 0, 1);
+    parent_.deliverHistogramToSinks(*this, value);
+    flags_ |= Flags::Used;
+  }
+
+  void beginMerge() { current_active_ = 1 - current_active_; }
+
+  void merge() override {}
+
+  bool used() const override { return flags_ & Flags::Used; }
+
+  const HistogramStatistics& intervalStatistics() const override { return *interval_statistics_; }
+
+  const HistogramStatistics& cumulativeStatistics() const override {
+    return *cumulative_statistics_;
+  }
+
+  void merge(histogram_t* target) {
+    std::unique_lock<std::mutex> lock(merge_lock_);
+    histogram_t* hist_array[1];
+    hist_array[0] = histograms_[1 - current_active_];
+    hist_accumulate(target, hist_array, ARRAY_SIZE(hist_array));
+    hist_clear(hist_array[0]);
+  }
+
+  Store& parent_;
+
+private:
+  int current_active_;
+  histogram_t* histograms_[2];
+  std::atomic<uint16_t> flags_;
+  std::shared_ptr<HistogramStatistics> interval_statistics_;
+  std::shared_ptr<HistogramStatistics> cumulative_statistics_;
+  mutable std::mutex merge_lock_;
+};
+
+typedef std::shared_ptr<ThreadLocalHistogramImpl> TlsHistogramSharedPtr;
+
+/**
+ * Log Linear Histogram implementation that is stored in the main thread.
+ */
+class HistogramParentImpl : public Histogram, public MetricImpl {
+public:
+  HistogramParentImpl(const std::string& name, Store& parent, std::string&& tag_extracted_name,
+                      std::vector<Tag>&& tags)
+      : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), parent_(parent) {
+    interval_histogram_ = hist_alloc();
+    cumulative_histogram_ = hist_alloc();
+    cumulative_statistics_.reset(new HistogramStatisticsImpl(cumulative_histogram_));
+    interval_statistics_.reset(new HistogramStatisticsImpl(interval_histogram_));
+  }
+
+  ~HistogramParentImpl() {
+    hist_free(interval_histogram_);
+    hist_free(cumulative_histogram_);
+  }
+
+  // Stats::Histogram
+  void recordValue(uint64_t) override {}
+
+  bool used() const override {
+    bool any_tls_used = false;
+    for (auto tls_histogram : tls_histograms_) {
+      if (tls_histogram->used()) {
+        any_tls_used = true;
+        break;
+      }
+    }
+    return any_tls_used;
+  }
+
+  void addTlsHistogram(TlsHistogramSharedPtr hist_ptr) { tls_histograms_.emplace_back(hist_ptr); }
+
+  /**
+   * This method is called during the main stats flush process for each of the histogram. This
+   * method iterates through the Tls histograms and collects the histogram data of all of them
+   * in to "interval_histogram_". Then the collected "interval_histogram_" is merged to a
+   * "cumulative_histogram". More details about threading model at
+   * https://github.com/envoyproxy/envoy/issues/1965#issuecomment-376672282.
+   */
+  void merge() override {
+    if (used()) {
```

Comment:
would be nice to have a `usedWorker()` which asserts lock is already acquired, then you can acquire lock once at top

Reply:
I did not get this comment. Can you elaborate more?

Reply:
My comment here is that you acquire the lock for `used()`, then release it, then reacquire it. A better model would be to do something like:

```
bool used() override {
  <lock>
  usedWorker();
}

void merge() override {
  <lock>
  if (usedWorker()) {
    ..
  }
}
```

You can avoid double acquiring